airflow/jobs/scheduler_job.py

```diff
@@ -1574,3 +1585,33 @@ def _cleanup_stale_dags(self, session: Session = NEW_SESSION) -> None:
             dag.is_active = False
             SerializedDagModel.remove_dag(dag_id=dag.dag_id, session=session)
         session.flush()
+
+    @provide_session
+    def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
+        """
+        Detects datasets that are no longer referenced in any DAG schedule parameters or task outlets and
+        sets the dataset is_orphaned flag to True
+        """
+        orphaned_dataset_query = (
+            session.query(DatasetModel)
+            .join(
+                DagScheduleDatasetReference,
+                isouter=True,
+            )
+            .join(
+                TaskOutletDatasetReference,
+                isouter=True,
+            )
+            # MSSQL doesn't like it when we select a column that we haven't grouped by. All other DBs let us
+            # group by id and select all columns.
+            .group_by(DatasetModel if session.get_bind().dialect.name == "mssql" else DatasetModel.id)
```

Comment:
Is there any benefit to not just doing `group_by(DatasetModel)` for all of the backends? That seems simpler and doesn't need a comment.

Reply:
The idealist in me thinks this is the Proper Way ™️ do do it, it's working, and I still think it's pretty clear what's going on (read: what SQL is being emitted).

The pragmatist in me recognizes that you're making a good point, but I think let's leave it unless or until it causes issues, then do that at that point.

How does that plan sound?

Reply:
Works for me. We can always tweak that later too.